pip/pip-330.md

```diff
@@ -0,0 +1,55 @@
+# PIP-330: getMessageById gets all messages
+
+# Motivation
+
+The `org.apache.pulsar.client.admin.Topics` provides `getMessageById(java.lang.String, long, long)` method to get the
+message, which returns one message. If the message id refers to a batch message, we can only get the first message, not
+all messages.
+
+This behavior affects our analysis of messages by the message id.
+
+# Goals
+
+## In Scope
+
+Add a method that returns all messages or specified message by message id to the `org.apache.pulsar.client.admin.Topics` interface.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. Add a set of methods to the `org.apache.pulsar.client.admin.Topics` interface:
+
+```java
+public interface Topics {
+    // When batchIndex is greater than or equal to 0, return the specified message.
+    List<Message<byte[]>> getMessageById(String topic, long ledgerId, long entryId, Long batchIndex) throws PulsarAdminException;
+    CompletableFuture<List<Message<byte[]>>> getMessageByIdAsync(String topic, long ledgerId, long entryId, Long batchIndex);
```

Comment:
```suggestion
    List<Message<byte[]>> getMessageById(String topic, long ledgerId, long entryId, int batchIndex) throws PulsarAdminException;
    CompletableFuture<List<Message<byte[]>>> getMessageByIdAsync(String topic, long ledgerId, long entryId, int batchIndex);
```

batch index should be a 32-bits integer.

In addition, you should document the behavior when `batchIndex` is negative.

Reply:
> batch index should be a 32-bits integer.
Or `Integet`? 

When `batchIndex` is null or negative, return all messages.

Reply:
This is a good idea. In most cases, we need to get all messages.